Moves Sync dialog outside of Unreal

kalisp opened the pull request:
New hook is created to show blocking dialog to sync specific change list before Unreal will launched.

It is expected that Perforce workspace is created/set and first version of Unreal project is synced (as discussed previously).

(Version-control logic was removed from `ayon-core` hook, `Sync To` button was removed from Unreal Ayon menu.)

Requires: 
- https://github.com/ynput/ayon-deadline/pull/1
- https://github.com/ynput/ayon-unreal/pull/8

antirotor reviewed (dismissed):
It cannot be easily tested without solving this:
https://github.com/ynput/ayon-version-control/blob/91e07fa19bbc40d8c5a0809aef821b4b2ebdc3e4/client/version_control/changes_viewer/control.py#L93-L97

kalisp commented:
> It cannot be easily tested without solving this:
> 
> https://github.com/ynput/ayon-version-control/blob/91e07fa19bbc40d8c5a0809aef821b4b2ebdc3e4/client/version_control/changes_viewer/control.py#L93-L97

Fixed.

kalisp commented:
> I think we should move it outside the pre-launch hook - to subprocess at least. Having it in pre-launch hook is troublesome because maybe in the future, we will need to run unreal headlessly for example.

This feature should be blocking, syncing needs to be done before Unreal editor start. Syncing on Deadline is being handled elsewhere.

antirotor commented:
not sure if this  shouldl be addressed in this PR, but dumping it here anyway:

![image](https://github.com/ynput/ayon-version-control/assets/33513211/3ff86d04-40b3-41d5-8990-ded988f51a75)

when you clikc "sync to" button, there is no visual feedback whatsoever - only log in the console if you have it opened. It has the feeling that it doesn't work where it in fact works. Should it (after successful sync) continue to Unreal?

antirotor commented on client/version_control/addon.py:
```diff
@@ -123,6 +123,16 @@ def get_publish_plugin_paths(self, host_name):
+        """Implementation for applications launch hooks.
+
+        Returns:
+            (str): full absolut path to directory with hooks for the module
```
```suggestion
            (str): full absolute path to directory with hooks for the module
```

antirotor reviewed (changes requested):
Just one typo and comment about product name, can be dismissed if not relevant. Otherwise it works.

antirotor commented:
Also - something was lost in the repository transfer. This states that it needs ynput/ayon-core#183 that was closed and replaced by ynput/ayon-core#670 that was closed yet again in favour of ynput/ayon-deadline#1 but I think something in 'ayon_core.modules.webserver` was changed, because it doesn't work with the latest develop version of ynput/ayon-core

<details>
<summary>Pretty hefty traceback at the start of ayon-launcher</summary>
<pre>
WARNING:TrayAddonsManager:Addon "version_control" crashed on `tray_start`.
Traceback (most recent call last):
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-core\client\ayon_core\tools\tray\addons_manager.py", line 202, in start_addons
    addon.tray_start()
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-version-control\client\version_control\addon.py", line 108, in tray_start
    from version_control.rest.communication_server import WebServer
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-launcher\vendor\python\shiboken2\files.dir\shibokensupport\__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-version-control\client\version_control\rest\communication_server.py", line 10, in <module>
    from version_control.rest.perforce.rest_api import PerforceModuleRestAPI
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-launcher\vendor\python\shiboken2\files.dir\shibokensupport\__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-version-control\client\version_control\rest\perforce\rest_api.py", line 4, in <module>
    from version_control.backends.perforce import rest_routes
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-launcher\vendor\python\shiboken2\files.dir\shibokensupport\__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-version-control\client\version_control\backends\perforce\__init__.py", line 38, in __getattr__
    return importlib.import_module(f"{__package__}.{name}")
  File "C:\Users\annat\.pyenv\pyenv-win\versions\3.9.13\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-version-control\client\version_control\backends\perforce\rest_routes.py", line 7, in <module>
    from ayon_core.modules.webserver.base_routes import RestApiEndpoint
  File "C:\Users\annat\Documents\Projects\Ayon\ayon-launcher\vendor\python\shiboken2\files.dir\shibokensupport\__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
ModuleNotFoundError: No module named 'ayon_core.modules.webserver.base_routes'
</pre>
</details>

kalisp commented on client/version_control/plugins/create/unreal/publish_commit.py:
```diff
@@ -38,27 +28,32 @@ def create(self, options=None):
-        if existing_instance is None:
-            asset_doc = get_asset_by_name(project_name, asset_name)
+
             product_name = self.get_product_name(
```
Renamed all occurences to `changelist_metadata`

kalisp commented:
> Also - something was lost in the repository transfer. This states that it needs [ynput/ayon-core#183](https://github.com/ynput/ayon-core/pull/183) that was closed and replaced by [ynput/ayon-core#670](https://github.com/ynput/ayon-core/pull/670) that was closed yet again in favour of [ynput/ayon-deadline#1](https://github.com/ynput/ayon-deadline/pull/1) but I think something in 'ayon_core.modules.webserver` was changed, because it doesn't work with the latest develop version of ynput/ayon-core
> 
> Pretty hefty traceback at the start of ayon-launcher

Updated imports.

antirotor reviewed (approved):
works, thanks!
